### ISSUE TITLE
Feature Request: Add support for CreateVPCAssociationAuthorization AWS API

### ISSUE THREAD

**Issue by hashibot** [terraform-providers/terraform-provider-aws#384]
_This issue was originally opened by @rginev as hashicorp/terraform#10208. It was migrated here as part of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original body of the issue is below._

<hr>

Hey there,

Terraform does have [route53_zone_association](https://www.terraform.io/docs/providers/aws/r/route53_zone_association.html) resource, but it works for private zones and VPCs in the same AWS account. 
If you want to associate a VPC that you created with one AWS account with a private hosted zone that you created with a different account you had to ask AWS support to create authorisation manually. 
Now they added an [API](http://docs.aws.amazon.com/Route53/latest/APIReference/API_CreateVPCAssociationAuthorization.html) for this. It would be great to have this implemented in terraform e.g. new resource.

Thanks in advance!

### References
- [Associating an Amazon VPC and a Private Hosted Zone That You Created with Different AWS Accounts](http://docs.aws.amazon.com/Route53/latest/DeveloperGuide/hosted-zone-private-associate-vpcs-different-accounts.html)

**Comment by josephholsten** [comment]
@RyanJarv I don't know if you've looked here since the provider split. Let me know if you want to help get a fix together for this.

**Comment by Zordrak** [comment]
Should only require the creation of an authorization resource against a specific zone, with parameter inputs for source vpc id and source vpc region.

Would love to see this implemented.

**Comment by RyanJarv** [comment]
@josephholsten Hey sorry, yeah that would be awesome I've just been pretty busy and not sure when I'll have time to get back to this.

Worth mentioning this actually does work currently for creating authorizations but like @Zordrak pointed out the association resource can't describe route53 zone and fails currently. This is talked about a bit more at https://github.com/hashicorp/terraform/pull/12553. If we don't mind the association and authorization resource working together currently I think this could probably be merged as is with some added docs and a note on the issue.

**Comment by robh007** [comment]
Is anyone still interested in trying to get this feature out?

**Comment by RyanJarv** [comment]
Went ahead and cleaned this up a bit and added documentation in https://github.com/terraform-providers/terraform-provider-aws/pull/2005. Last PR was before the repo split.

**Comment by julianalves** [comment]
Awesome! this will be very useful! Thanks for the hard work!

**Comment by luisamador** [comment]
Agreed. Fixing this would be very helpful. In the meantime I have used the AWS CLI to do the Route53 VPC association from a VPC sitting in one AWS account to a Route53 Hosted Zone sitting on different AWS account. Here are the steps: https://aws.amazon.com/premiumsupport/knowledge-center/private-hosted-zone-different-account/

**Comment by ewbankkit** [comment]
There's a problem with the way the Route 53 API works when creating the association between a hosted zone created in one AWS account and a VPC created in a second AWS account.

This Terraform code (following the steps [here](https://aws.amazon.com/premiumsupport/knowledge-center/private-hosted-zone-different-account/), and based on [this code](https://github.com/ewbankkit/terraform-provider-aws/tree/issue-384))
```hcl
provider "aws" {
  // Zone creator's credentials.
}

provider "aws" {
  alias = "bar"
  // VPC creator's credentials.
  access_key = "XXXXXXXXXXXXXXXXXXXX"
  secret_key = "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
}

resource "aws_vpc" "foo" {
  cidr_block           = "10.6.0.0/16"
  enable_dns_hostnames = true
  enable_dns_support   = true
}

resource "aws_route53_zone" "foo" {
  name   = "example.com"
  vpc_id = "${aws_vpc.foo.id}"
}

resource "aws_vpc" "bar" {
  provider = "aws.bar"

  cidr_block           = "10.7.0.0/16"
  enable_dns_hostnames = true
  enable_dns_support   = true
}

resource "aws_route53_zone_association_authorization" "foo" {
  zone_id = "${aws_route53_zone.foo.id}"
  vpc_id  = "${aws_vpc.bar.id}"
}

resource "aws_route53_zone_association" "bar" {
  provider = "aws.bar"

  zone_id = "${aws_route53_zone_association_authorization.foo.zone_id}"
  vpc_id  = "${aws_vpc.bar.id}"
}
```
should create VPCs in two different AWS accounts and associate them both with a hosted zone created in the first AWS account. In fact it does exactly that, but an error is reported:
```
terraform apply
aws_vpc.bar: Creating...
  assign_generated_ipv6_cidr_block: "" => "false"
  cidr_block:                       "" => "10.7.0.0/16"
  default_network_acl_id:           "" => "<computed>"
  default_route_table_id:           "" => "<computed>"
  default_security_group_id:        "" => "<computed>"
  dhcp_options_id:                  "" => "<computed>"
  enable_classiclink:               "" => "<computed>"
  enable_classiclink_dns_support:   "" => "<computed>"
  enable_dns_hostnames:             "" => "true"
  enable_dns_support:               "" => "true"
  instance_tenancy:                 "" => "<computed>"
  ipv6_association_id:              "" => "<computed>"
  ipv6_cidr_block:                  "" => "<computed>"
  main_route_table_id:              "" => "<computed>"
aws_vpc.foo: Creating...
  assign_generated_ipv6_cidr_block: "" => "false"
  cidr_block:                       "" => "10.6.0.0/16"
  default_network_acl_id:           "" => "<computed>"
  default_route_table_id:           "" => "<computed>"
  default_security_group_id:        "" => "<computed>"
  dhcp_options_id:                  "" => "<computed>"
  enable_classiclink:               "" => "<computed>"
  enable_classiclink_dns_support:   "" => "<computed>"
  enable_dns_hostnames:             "" => "true"
  enable_dns_support:               "" => "true"
  instance_tenancy:                 "" => "<computed>"
  ipv6_association_id:              "" => "<computed>"
  ipv6_cidr_block:                  "" => "<computed>"
  main_route_table_id:              "" => "<computed>"
aws_vpc.bar: Creation complete after 6s (ID: vpc-951bcaec)
aws_vpc.foo: Creation complete after 5s (ID: vpc-b018c9c9)
aws_route53_zone.foo: Creating...
  comment:        "" => "Managed by Terraform"
  force_destroy:  "" => "false"
  name:           "" => "example.com"
  name_servers.#: "" => "<computed>"
  vpc_id:         "" => "vpc-b018c9c9"
  vpc_region:     "" => "<computed>"
  zone_id:        "" => "<computed>"
aws_route53_zone.foo: Still creating... (10s elapsed)
aws_route53_zone.foo: Still creating... (20s elapsed)
aws_route53_zone.foo: Still creating... (30s elapsed)
aws_route53_zone.foo: Creation complete after 35s (ID: Z3UMIQBZQ9Y13C)
aws_route53_zone_association_authorization.foo: Creating...
  vpc_id:     "" => "vpc-951bcaec"
  vpc_region: "" => "<computed>"
  zone_id:    "" => "Z3UMIQBZQ9Y13C"
aws_route53_zone_association_authorization.foo: Creation complete after 0s (ID: Z3UMIQBZQ9Y13C:vpc-951bcaec)
aws_route53_zone_association.bar: Creating...
  vpc_id:     "" => "vpc-951bcaec"
  vpc_region: "" => "<computed>"
  zone_id:    "" => "Z3UMIQBZQ9Y13C"
aws_route53_zone_association.bar: Still creating... (10s elapsed)
aws_route53_zone_association.bar: Still creating... (20s elapsed)
aws_route53_zone_association.bar: Still creating... (30s elapsed)
Error applying plan:

1 error(s) occurred:

* aws_route53_zone_association.bar: 1 error(s) occurred:

* aws_route53_zone_association.bar: AccessDenied: User: arn:aws:iam::000000000000:user/kit is not authorized to access this resource
	status code: 403, request id: c1f53cc1-f6fe-11e7-8456-41d8d6fcb1f8

Terraform does not automatically rollback in the face of errors.
Instead, your Terraform state file has been partially updated with
any resources that successfully completed. Please address the error
above and apply again to incrementally change your infrastructure.
```
and subsequent `terraform plan`s get the same error:
```
terraform plan
Refreshing Terraform state in-memory prior to plan...
The refreshed state will be used to calculate this plan, but will not be
persisted to local or remote state storage.

aws_vpc.bar: Refreshing state... (ID: vpc-951bcaec)
aws_vpc.foo: Refreshing state... (ID: vpc-b018c9c9)
aws_route53_zone.foo: Refreshing state... (ID: Z3UMIQBZQ9Y13C)
aws_route53_zone_association_authorization.foo: Refreshing state... (ID: Z3UMIQBZQ9Y13C:vpc-951bcaec)
aws_route53_zone_association.bar: Refreshing state... (ID: Z3UMIQBZQ9Y13C:vpc-951bcaec)
Error refreshing state: 1 error(s) occurred:

* aws_route53_zone_association.bar: 1 error(s) occurred:

* aws_route53_zone_association.bar: aws_route53_zone_association.bar: AccessDenied: User: arn:aws:iam::000000000000:user/kit is not authorized to access this resource
	status code: 403, request id: 22ab1eca-f700-11e7-aa35-9bd7d0c4ca7d
```

The problem is with the code that run under the credentials of the associator (i.e. NOT the AWS account that created the hosted zone) that is reading the state of the zone association https://github.com/terraform-providers/terraform-provider-aws/blob/17a320b90f85a7e98e19f2b405d0526b640f4880/aws/resource_aws_route53_zone_association.go#L94-L95
As there is no AWS zone associated resource the zone itself is queried and this fails with the _AccessDenied_ error when called from the AWS account that didn't create the zone.

**Comment by RyanJarv** [comment]
@ewbankkit Yeah this is an issue with  aws_route53_zone_association and is documented in my PR at https://github.com/terraform-providers/terraform-provider-aws/pull/2005/files#diff-f048cbf19e0fc509e5df5d05254f5bdbR13.

This was also the reason I held off on making that PR for a bit, some people where asking for it regardless so short of a better fix I added a note on the issue (although looking at it now this should be added to the `aws_route53_zone_association` doc as well). I'm sure this could be worked around somehow, really I just come up with anything didn't feel hacky to me at the time.

**Comment by JoshiiSinfield** [comment]
@RyanJarv 
Hi, have you considered requiring a second provider for which to do the R53 api call with?

If it's not a TF coding anti-pattern / no-no I'd be happy with having to specifiy two providers to a resource. I'd expect the authorisation to be created at the same time as the association as the example above does, so the second provider should be available anyway.

A second idea I've had, but the answer doesn't seem obvious from the AssociateVPCWithHostedZone API reference, What error/response do you get when trying to associate an already associated VPC to a zone? Is the response suitable to be used in determining whether the VPC is already associated, rather than having to get the list of vpcs from the zone?

excuse my naivety if these are obvious no's! 😄 

Cheers, 
Josh

**Comment by psyvision** [comment]
@JoshiiSinfield If I'm reading the above correctly then @ewbankkit has already tried doing this (hence the two `provider "aws"` statements at the top of his TF code.

From using the aws cli, from memory, I think you do get a different message when the zone has already been associated.

**Comment by JoshiiSinfield** [comment]
@psyvision 
Yes he has, except I was on about passing a second provider into the zone_association resource. 

Something like this (assuming use of the rest of @ewbankkit's code)::

```
resource "aws_route53_zone_association" "bar" {
  provider = "aws.bar"
  r53_zone_provider = "aws"

  zone_id = "${aws_route53_zone_association_authorization.foo.zone_id}"
  vpc_id  = "${aws_vpc.bar.id}"
}
```

**Comment by RyanJarv** [comment]
I think I remember considering something like that at one point,
unsure if I just didn’t think it was possible or wasn’t sure how to
implement it.
If it’s doable though seems like a pretty good work around. But I might
need to take a step back from this PR and would be great if some one is
able to take the lead here.

On Mon, Mar 12, 2018, at 4:19 AM, Josh Sinfiel
> 


> @psyvision[1]
> Yes he has, except I was on about passing a second provider into the
> zone_association resource.> 


> Something like this (assuming use of the rest of
> @ewbankkit[2]'s code)::


> resource "aws_route53_zone_association" "bar" { provider = "aws.bar"
> r53_zone_provider = "aws"  zone_id =
> "${aws_route53_zone_association_authorization.foo.zone_id}" vpc_id  =
> "${aws_vpc.bar.id}" }
>> — You are receiving this because you were mentioned. Reply to this
> email directly, view it on GitHub[3], or mute the thread[4].> 


Links:

  1. https://github.com/psyvision
  2. https://github.com/ewbankkit
  3. https://github.com/terraform-providers/terraform-provider-aws/issues/384#issuecomment-372276327

**Comment by tdmalone** [comment]
Related: #585 and #8593

**Comment by RyanJarv** [comment]
Hey all, haven't got around to looking into this again yet but did want to leave this here, turn's out the API issue thing is also a bit of a security issue. I reported this to AWS security about two years ago when I was working on this originally but am unsure of any timeline on fixing it.

https://blog.ryanjarv.sh/2019/05/24/backdooring-route53-with-cross-account-dns.html

**Comment by aeschright** [comment]
Hi everyone, we've decided that there isn't a way we're happy to do this currently within the API constraints. Please don't submit new PRs for the resource right now. I'm going to leave the ticket open as an upstream issue to reconsider if something changes.

For a workaround, you can use the cli instructions instead: https://aws.amazon.com/premiumsupport/knowledge-center/private-hosted-zone-different-account/

**Comment by mhaley-miovision** [comment]
Hi, I see that you guys don't want to go forward with this PR because the API does not have a good way of allowing an account to view vpc associations on a route53 zone that it does not own.  I think that there are still uses for this resource and see no reason why limitation in associations API should stop you from implementing the authorizations API.  All the pieces are there.  And who knows, maybe once people get their hands on this, the path to full cross account route53 associations managed by terraform will be more clear.

**Comment by krogon** [comment]
Hi, I was able to workaround this limitations using multiple data sources and resource. Please let me know if this would be accepted, as then I would need additional time for testing coverage.

**Comment by RyanJarv** [comment]
@krogon the issue could have been worked around before, the problem has always been it's hacky at best. If you think you have a good solution and have the code available I'd be happy to take a look at it.

**Comment by ipnextgen** [comment]
Why not simply split it into two ressources with a delay feature? Would then be easy to specify a provider for the auth ressource and another for the association ressource? Its just lame we still need to do some CLI calls when 99% of the infrastructure is all in TF!!!

**Comment by mhaley-miovision** [comment]
@ipnextgen my workaround for these types of things right now has been to use https://github.com/scottwinkler/terraform-provider-shell.

The terraform ends up being very simple. Here is an example for security hub invitations.  Basically you invoke the resource and tell it what script to call.
```
resource "shell_script" "security_hub_membership" {
  lifecycle_commands {
    create = "./scripts/security_hub_invitation.sh create ${data.aws_caller_identity.security.account_id} ${local.account_info.id} ${local.account_info.email}"
    read   = "./scripts/security_hub_invitation.sh read ${data.aws_caller_identity.security.account_id} ${local.account_info.id}"
    delete = "./scripts/security_hub_invitation.sh delete ${data.aws_caller_identity.security.account_id} ${local.account_info.id}"
  }

  working_directory = path.module
}
```

Then I have structured all my scripts like so
```
#!/usr/bin/env bash
set -e -x -o pipefail

## Goal: Provide a script that terraform shell provider can use to manage security hub invitations.

## Helpful function to assume a role and run an AWS command
## arguments:
## 1 -> account id
## 2..n -> aws cli arguments
function AWS_RUN_CMD_AS {
    ##Get temporary credentials for the account we were given
    CREDENTIALS=$(aws sts assume-role --role-arn arn:aws:iam::$1:role/YourRoleName \
        --role-session-name enable-security-hub)

    ##Setup the proper temp credentials and invoke the aws cli with the provided args
    AWS_ACCESS_KEY_ID=$(echo $CREDENTIALS | jq -r '.Credentials.AccessKeyId') \
    AWS_SECRET_ACCESS_KEY=$(echo $CREDENTIALS | jq -r '.Credentials.SecretAccessKey') \
    AWS_SESSION_TOKEN=$(echo $CREDENTIALS | jq -r '.Credentials.SessionToken') \
    aws "${@:2}"
}

## arguments:
## 1 -> Security account id
## 2 -> id of account to create membership for
## 3 -> email of account to create membership for
function create {
  ## implement create function
}

## arguments:
## 1 -> Security account id
## 2 -> account id to check status for
function read {
  ## implement read function
}

function update {
  echo "Update is not implemented."
  exit 6
}

## arguments:
## 1 -> Security account id
## 2 -> account id to delete membership
function delete {
  ##Implement delete function
}

case $1 in
  create)
    create "${@:2}"
    ;;
  read)
    read "${@:2}"
    ;;
  update)
    update "${@:2}"
    ;;
  delete)
    delete "${@:2}"
    ;;
  *)
    echo "Unrecognized command $1"
    exit 5
    ;;
esac
```

I don't like to use the shell provider too much, but the main benefit here is that we get the normal CRUD operations.  This allows us to do normal terraform things, whereas using the built in local-exec really only lets you create and delete (sort of).

Creating a script to run the CLI commands to create/read/delete route53 authorizations should be pretty straight forward.  The json output from the read operation is made available in terraform like any other resource so this should fit pretty seamlessly into your existing setup.

**Comment by RyanJarv** [comment]
@ipnextgen you are misunderstanding the issue.

**Comment by miketwenty1** [comment]
@RyanJarv 
I just opened this up [here](https://github.com/hashicorp/terraform/issues/24286) #24286 in terraform repo, but didn't realize this was considered a provider issue?

I would imagine this just needs 2 resources similar to how to setting up a peer connection works. 
respectfully:
aws_vpc_peering_connection
aws_vpc_peering_connection_accepter

Maybe we would need:
aws_route53_peer_association
aws_route53_peer_association_accepter

This is currently how you do this with the API manually.
## APPLY
```sh
aws route53 create-vpc-association-authorization \
--hosted-zone-id ZAZAPAPAXAXAXAXAXA10 \
--vpc VPCRegion=us-east-1,VPCId=vpc-000bb00bb00bb00bb \
--profile profile_with_access_to_private_hosted_zone
```
```sh
aws route53 associate-vpc-with-hosted-zone \
--hosted-zone-id ZAZAPAPAXAXAXAXAXA10 \
--vpc VPCRegion=us-east-1,VPCId=vpc-000bb00bb00bb00bb \
--profile profile_with_access_to_vpc_in_secondary_account
```

## DESTROY
```sh
aws route53 delete-vpc-association-authorization \
--hosted-zone-id ZAZAPAPAXAXAXAXAXA10 \
--vpc VPCRegion=us-east-1,VPCId=vpc-000bb00bb00bb00bb \
--profile profile_with_access_to_private_hosted_zone
```
```sh
aws route53 disassociate-vpc-from-hosted-zone \
--hosted-zone-id ZAZAPAPAXAXAXAXAXA10 \
--vpc VPCRegion=us-east-1,VPCId=vpc-000bb00bb00bb00bb \
--profile profile_with_access_to_vpc_in_secondary_account
```

**Comment by RyanJarv** [comment]
> @RyanJarv does any of this change your opinion on how this would be tackled?

I kinda forget what the issue is exactly every time I look at this post, so probably worth taking whatever I say here with a grain of salt.

The association needs to be run in the hosted zone account iirc, so the error is because this account can’t enumerate current associations. It can work but you start to run into cases where you don’t know the current state of stuff. It seems like this is still an issue I believe? If you wanted to only store state about associations locally that seems doable, I’m not sure I understand the impact of doing this though.

**Comment by miketwenty1** [comment]
@RyanJarv 
> So if you take out the read part of the CRUD stuff on the association resource this should work(?)

Are addressing the 403 error I put above? I believe so.

> If the resource is ever managed outside of terraform you will run into issues though.

We are strictly speaking about just association of VPC-Account B to private hosted zone-Account A.. i think this is ok.

> if you could somehow grant access to read the zone and maybe include it in an example of how 
this should be used maybe that would work as well.

Without the idea of assuming more than 1 profile/alias within a resource/module (role-ception) this doesn't seem doable with terraform/aws. I question though how the above command works fine though specifying only one profile. Route53 and VPC's don't have the capability of a s3 bucket policy equivalent where you can grant access to IAM users/roles in different accounts.

```
aws route53 associate-vpc-with-hosted-zone \
--hosted-zone-id ZAZAPAPAXAXAXAXAXA10 \
--vpc VPCRegion=us-east-1,VPCId=vpc-000bb00bb00bb00bb \
--profile profile_with_access_to_vpc_in_secondary_account
```

I mainly point to the cli commands that I reference above. Both manual cases for applying and destroying work perfect with just 1 profile being used with each command. I would imagine this translates to the API.. given it only has the inputs I provide.
When running the associate cli command again after it's already associated. I get this.

`
An error occurred (ConflictingDomainExists) when calling the AssociateVPCWithHostedZone operation: The VPC that you chose, vpc-00000000000 in region us-east-1, is already associated with another private hosted zone that has an overlapping name space, somedomain.com..
`

Meaning it does plenty fine knowing it's already associated. This coming from the profile of the account with the VPC that doesn't have direct access to the private hosted zone in the "route53" account.

> The other part of this is AWS recommends removing the authorization after the association is complete. But tbh I’m not really sure I understand the reasoning for doing this, or why you wouldn’t want a VPC to be re-associated. 

This is because after it is associated you no longer need it authorized. I think like you mention this can be noted in the docs.. and probably ignored.

**Comment by miketwenty1** [comment]
I would like to point out.

Seems that this was first reported by @rginev on Nov 17, 2016 (roughly ~3.5 years ago)
https://github.com/hashicorp/terraform/issues/10208

Since then:

https://github.com/hashicorp/terraform/issues/11174 (closed -> ^10208)
https://github.com/hashicorp/terraform/issues/12804 (closed -> #617)
https://github.com/hashicorp/terraform/issues/20841 (closed no resolution)
https://github.com/hashicorp/terraform/issues/22780 (closed -> #9453)
https://github.com/hashicorp/terraform/issues/24286 (closed -> #384 here)

https://github.com/hashicorp/terraform/pull/12553 (@RyanJarv your PR)

https://github.com/terraform-providers/terraform-provider-aws/issues/384 (here)
https://github.com/terraform-providers/terraform-provider-aws/issues/585 (open)
https://github.com/terraform-providers/terraform-provider-aws/issues/617 (open)
https://github.com/terraform-providers/terraform-provider-aws/pull/2005 (closed no resolution)
https://github.com/terraform-providers/terraform-provider-aws/issues/7805 (closed no resolution)
https://github.com/terraform-providers/terraform-provider-aws/issues/7812 (closed -> #7805)
https://github.com/terraform-providers/terraform-provider-aws/issues/8593 (open)
https://github.com/terraform-providers/terraform-provider-aws/issues/9453 (open)

There's probably more I missed, I'm just trying to point out that this has been wanted for a while.

Both @ewbankkit @FransUrbo seem to have a lot of history with this and might be interested in this still? 
@bflad (maintainer?)
This also came up in some research:
https://forums.aws.amazon.com/thread.jspa?threadID=271250

**Comment by ewbankkit** [comment]
I thought a lot about this a while back and my conclusion was that capturing the behaviour of this API in a resource doesn't match the way that the AWS provider works - i.e. the interactions for a single resource are done under the single AWS identity configured in the resource.
One solution may be to create separate provider that contains this new resource and this new provider is configured with both AWS identities. The complexity would all be in the myriad options for configuring those identities for the underlying AWS SDK.
https://github.com/hashicorp/aws-sdk-go-base could contain some reusable parts.

**Comment by miketwenty1** [comment]
@ewbankkit 
> the interactions for a single resource are done under the single AWS identity configured in the resource.

I'm not understanding the issue tbh. 
looking over:
[API_CreateVPCAssociationAuthorization](https://docs.aws.amazon.com/Route53/latest/APIReference/API_CreateVPCAssociationAuthorization.html)
and
[API_AssociateVPCWithHostedZone](https://docs.aws.amazon.com/Route53/latest/APIReference/API_AssociateVPCWithHostedZone.html)

Seems that this would only require a single aws identity per resource/api call/cli cmd.
I quote from the 2nd link.
```
Use the AWS account that created the private hosted zone to submit a 
CreateVPCAssociationAuthorization request. Then use the account that created the VPC to submit
 an AssociateVPCWithHostedZone request.

If a subnet in the VPC was shared with another account, you can use the account that the subnet 
was shared with to submit an AssociateVPCWithHostedZone request. For more information about 
sharing subnets, see Working with Shared VPCs.
```

Sounds like it follows suit with the commands I posted above. https://github.com/terraform-providers/terraform-provider-aws/issues/384#issuecomment-595276897

Would you be able to explain what the issue is exactly?

**Comment by miketwenty1** [comment]
wanted to bump you on this @ewbankkit and maybe also @RyanJarv 
Any thoughts on?

**Comment by RyanJarv** [comment]
Is anyone aware of any examples in terraform where the read part of CRUD is skipped, leaving other operations to depend only the data in the state file?

**Comment by evandam** [comment]
I'm running into the same issue and would love to see a solution, or at least some extra documentation that outlines this scenario before more people end up running it and finding this issue.

**Comment by wstewartii** [comment]
I'm running into this issue as well and I think anyone trying to make use of a private hosted zone across accounts with terraform is going to run into this issue. Managing multiple aws accounts is a very common use case in large companies and is what aws seems to recommend as an architecture within an organization. It's very disappointing to see this go unaddressed for years.

**Comment by tony-kerz** [comment]
just encountering this issue myself.

there are ways to address using resolver-endpoints (inbound at _phz-provider_, outbound at _phz-consumer_), but those endpoints are charged separately by aws and it would be great to avoid that extra charge with the "direct" phz <-> vpc association alternative.

per [this knowledge-base article](https://aws.amazon.com/premiumsupport/knowledge-center/private-hosted-zone-different-account/) it shows how to accomplish the authorization via cli.

i can/will try myself, but figured i would pose the question to the experienced participants in this thread: 

assuming that the __authorization__ part was done manually in the context of the _phz-provider_ account, would use of [aws_route53_zone_association](https://www.terraform.io/docs/providers/aws/r/route53_zone_association.html) work in the context of the _phz-consumer_ account?

**Comment by RyanJarv** [comment]
Hi all, I was looking into this again after @miketwenty1 pointed out that you can get the state of the `aws_route53_zone_association` from the error messages returned after repeated calls to that API.

I put something together to test this out in https://github.com/terraform-providers/terraform-provider-aws/compare/master...RyanJarv:cross-account-hostedzone-association.

This is all just a test right now, so any thoughts here are appreciated. In that branch I'm using a cross_account attribute on the `aws_route53_zone_association` resource to change the behavior of how the read and wait parts are done.

So in that branch if `cross_account = true`:
* We only rely on local state for the Read function
* Instead of calling get-wait to determine when Create is done we do repeated calls to AssociateVPCWithHostedZone and wait for the error message to inform us that this zone is already associated.

This seems to work ok but at the moment but has the following issues:
* Not being able to detect drift of the `aws_route53_zone_association` resource.
* The workflow for association authorization still won't follow step 7 here https://aws.amazon.com/premiumsupport/knowledge-center/private-hosted-zone-different-account/.
 
I don't think the second is really an issue but wanted to mention it since it was brought up before. For the first there might be a few ways to handle this but I still need to look into how practical they are.

Mostly wanted to post this now to see if anyone has any opinions on this behavior.

**Comment by RyanJarv** [comment]
@evandam
> Alternatively, would it be possible to have an IAM policy that would allow the account with the VPC to list hosted zones on the account with the hosted zone as a workaround?

Using assumeRole yes, however terraform afik isn't designed to connect to multiple accounts in a single resource. If it was possible to do this without assumeRole this would be more feasible. I wasn't able to figure either of these options out though at least in a way that made sense.

@tony-kerz 
> assuming that the authorization part was done manually in the context of the phz-provider account, would use of aws_route53_zone_association work in the context of the phz-consumer account?

This would not work, for the most part aws_route53_zone_association is the problem. If you want to try the branch in my last post that may work for you (warning: have not done much testing on it).

**Comment by hamishfor** [comment]
We have encountered this issue as well. Trying to build an ephemeral organization that has DNS is turning out to be quite the challenge.

This... _bug_? Will become more and more pertinent to businesses that have spawned in the era of IAC as they grow to incorporate multi account architectures, at AWS's recommendation I might add!

**Comment by avvi00** [comment]
I'm also keen to see a clean resolution to this issue and I understand this is a complex one to solve!

I've tried using a null_resource to call the CLI and do the association. The problem is that when you create the zone, you can't specify the other account VPC or you get an error, so after everything has run the first time and the VPC is associated with the zone, terraforms seeks to remove the VPC from the hosted zone. It's all pretty ugly to handle at the moment.

**Comment by tdmalone** [comment]
https://aws.amazon.com/about-aws/whats-new/2020/06/amazon-route-53-launches-new-api-action-to-list-private-hosted-zones-associated-with-your-amazon-vpcs 🎉

**Comment by ewbankkit** [comment]
Requires [AWS SDK v1.32.5](https://github.com/aws/aws-sdk-go/releases/tag/v1.32.5):
* https://github.com/terraform-providers/terraform-provider-aws/pull/13803

**Comment by AzySir** [comment]
@tdmalone great timing with that post - has someone taken ownership to bring this home?

**Comment by arnvid** [comment]
Any chance this will get added into the provider anytime soon? The workarounds given uses local exec for aws cli and this is something we are trying very hard to avoid when using Terraform Enterprise. Also we are going into 100+ accounts now - and the less dependencies on externals the better :)

**Comment by tdmalone** [comment]
@arnvid Copying an update from Brian on the above PR:
> Marking this for right after we complete our work for 3.0 in the next week or two, although review cycles might make it slightly after 3.1 for actual release. 👍

So, this is looking promising for pretty soon! 🎉

**Comment by bflad** [comment]
Support for Route 53 cross-account VPC Associations via a new `aws_route53_vpc_association_authorization` resource and updated `aws_route53_zone_association` resource handling has been merged and will release with version 3.1.0 of the Terraform AWS Provider, likely later today. Thanks to @goakley and @RyanJarv for implementation. 👍
